### ISSUE TITLE
Questions about demo_optimizer.py

### ISSUE THREAD

**Issue by callmexg1** [edoRemelli/MeshSDF#1]
In demo_optimizer.py:189
https://github.com/cvlab-epfl/MeshSDF/blob/7c169600872f7d77c2f3782ba03c7f87377100c3/demo_optimizer.py#L189
Why the loss is calculated by ``torch.sum(pred_sdf)`` instead of ``L(pred_sdf, gt_sdf)``

In addition, it seems that the main training script does not adopt Theorem 1 in the paper... Its optimization process is barely the same as DeepSDF?

**Comment by bguillard** [comment]
Hi,
In the following group of instructions:
https://github.com/cvlab-epfl/MeshSDF/blob/7c169600872f7d77c2f3782ba03c7f87377100c3/demo_optimizer.py#L188-L192 what actually happens is that we compute the surface normals of 3D points xyz (which we assume lie on the surface). This is step 4 of Algorithm 2 in the main paper. Indeed, geometry gives us that on the surface the gradient of the SDF field equates the surface normals: 
<img width="200" alt="Screenshot 2020-11-03 at 16 25 06" src="https://user-images.githubusercontent.com/23718226/98005296-2b8e4500-1df1-11eb-8cd5-697d7392fdfd.png">

More precisely, to do so we first compute the respective SDF values of the points (this is ```pred_sdf = decoder()```), then automatically differentiate it w.r.t. to their 3D position (doing ```sum()``` then ```backward()```). As a result, normals of xyz points are then in ```xyz.grad```, which we store in ```normals```.


For your second point: this is right, as the name suggests in ```train_deep_sdf.py``` we're only training a DeepSDF network according to the original procedure. We do this to get a net and codes for representing a dataset of two shapes, bob, a genus-1 rubber duck, and spot, a genus-0 cow. This trained net (+ codes) are then used in ```demo_optimizer.py``` to demonstrate we can backprop from pixel values all the way back to a latent code, via an intermediate explicit mesh.

I hope this clarifies your issues.

**Comment by callmexg1** [comment]
Thanks a lot! I think I misunderstand the algorithm. Issue closed.
